Upgrade to DLPack 1.0.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150691
* #150218
* #150217
* #150216
* __->__ #145000

This PR makes the necessary changes in order to upgrade PyTorch DLPack
support to version 1.0. In summary, we add support for the following:

- Support both `DLManagedTensor` and `DLManagedTensorVersioned` when
  producing and consuming DLPack capsules
- New parameter for `__dlpack__` method: `max_version`
- Version checks:
    - Fallback to old implementation if no `max_version` or if version
      lower than 1.0
    - Check that the to-be-consumed capsule is of version up to 1.X

In order to accommodate these new specifications, this PR adds the
following main changes:

- `torch._C._to_dlpack_versioned` Python API (Module.cpp): new Python
API for creating a versioned DLPack capsule (called by `__dlpack__`
method)
- `DLPackTraits<T>` class (DLConvertor.h): select the correct
traits (e.g. capsule name, conversion functions) depending on which 
DLPack tensor class is being used
- `toDLPackImpl<T>` function (DLConvertor.cpp): populates the
common fields of both classes
- `fromDLPackImpl<T>` function (DLConvertor.cpp): constructs a tensor
from a DLPAck capsule
- `fillVersion<T>` function (DLConvertor.cpp): populates the version
field for `DLManagedTensorVersioned` (no-op for `DLManagedTensor`)
- `tensor_fromDLPackImpl<T>` function (tensor_new.cpp): outer function
for constructing a tensor out of a DLPack capsule that also marks the
capsule as used

cc @ezyang @gchanan

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

albanD commented on aten/src/ATen/DLConvertor.h:
```diff
@@ -11,11 +11,50 @@
-TORCH_API Tensor fromDLPack(DLManagedTensor* src);
-TORCH_API Tensor
-fromDLPack(DLManagedTensor* src, std::function<void(void*)> deleter);
+TORCH_API DLManagedTensorVersioned* toDLPack(const Tensor& src);
```
Is this API used by C++ libraries? This would be a BC-breaking change for these users right?

ysiraichi replied:
I think it's being used by PyTorch/XLA. Yes, it's definitely BC-breaking. I was thinking that, since DLPack 1.0 should be the new default, the old version should have the name with a suffix. However, now that you brought this up, not being BC-breaking sounds more important.

In summary, I will change the names so that we are not BC-breaking.

albanD commented on aten/src/ATen/DLConvertor.h:
```diff
@@ -11,11 +11,50 @@
+TORCH_API DLManagedTensor* toDLPackUnversioned(const Tensor& src);
+TORCH_API Tensor fromDLPack(
+    DLManagedTensorVersioned* src,
+    std::optional<std::function<void(void*)>> deleter = std::nullopt);
```
std::function is optional by default and doesn't require std::optional right?

albanD commented on aten/src/ATen/dlpack.h:
```diff
@@ -15,11 +15,11 @@
+#define DLPACK_MAJOR_VERSION 1
 
-/*! \brief The current ABI version of dlpack */
-#define DLPACK_ABI_VERSION 1
```
The is no more ABI version?

albanD commented on torch/__init__.py:
```diff
@@ -2270,7 +2270,7 @@ def compiled_with_cxx11_abi() -> builtins.bool:
     solve,
 )
-from torch.utils.dlpack import from_dlpack, to_dlpack
+from torch.utils.dlpack import from_dlpack, to_dlpack, to_dlpack_unversioned
```
While I understand this matches current code, I don't think we want to have this as a new top level API? Being part of torch.utils.dlpack is enough?

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -1599,6 +1616,7 @@ static std::initializer_list<PyMethodDef> TorchMethods = {
      METH_NOARGS,
      nullptr},
     {"_to_dlpack", THPModule_toDLPack, METH_O, nullptr},
+    {"_to_dlpack_unversioned", THPModule_toDLPackUnversioned, METH_O, nullptr},
```
Why do we still need this one?

rgommers replied:
This does look necessary to me. For DLPack to change the ABI once, there's a dance that needs doing to be not-super-disruptive: continue returning the old (0.8) version, unless the consumer indicates it can handle the new (1.X) version by passing in `max_version=(1, 0)` (or `(1, x)` in the future).

albanD commented on torch/utils/dlpack.py:
```diff
@@ -98,23 +111,33 @@ def from_dlpack(ext_tensor: Any) -> 'torch.Tensor':
+        try:
+            # Try running __dlpack__ while specifying `max_version` argument.
+            dlpack = ext_tensor.__dlpack__(**kwargs)
+        except TypeError:
```
Is it guaranteed that they will fail with TypeError?

albanD reviewed (commented):
Change sounds ok but what is the expectations in terms of BC when interracting with libraries that didn't upgrade to latest dlpack yet? Is that ok for all Tensors to be of the new version?

rgommers commented on torch/_tensor.py:
```diff
@@ -1661,7 +1661,7 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
     __torch_dispatch__ = _C._disabled_torch_dispatch_impl
 
-    def __dlpack__(self, stream=None):
+    def __dlpack__(self, stream=None, max_version=None):
```
Note that there are also new `dl_device` and `copy` keywords ([API docs](https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__dlpack__.html)).

rgommers reviewed (commented):
Thanks for tackling this @ysiraichi! This looks like a good start. A few high-level comments:
- As noted inline, `__dlpack__` gained 3 new keywords, so far this adds only `max_version`. I suspect it's safer to add all at once, because other libraries are probably going to assume that if `max_version` is present, the 1.0 support is complete.
- It would be good to have new Python-level tests in [test/test_dlpack.py](https://github.com/pytorch/pytorch/blob/e9f6e273e7c517021c45090de795ad4e9016ad25/test/test_dlpack.py). That will also make the changes in logic easier to review.
    - There's one testing TODO for a very old numpy version there that may be nice to take along: https://github.com/pytorch/pytorch/blob/e9f6e273e7c517021c45090de795ad4e9016ad25/test/test_dlpack.py#L239
- I think this is inactionable right now, but adding for completeness: DLPack gained a new `DLPACK_FLAG_BITMASK_READ_ONLY` field, which in the future can feed into PyTorch's copy-on-write (COW) feature.

ysiraichi commented:
@albanD @rgommers Since you have already reviewed this PR, I was thinking about adding the new `__dlpack__` keywords + tests in a new PR stacked on top of this one. There are 2 reasons for that:

- It would be easier to review the new changes
- We could merge the whole stack at once when that PR is ready

Let me know if that works for you.

rgommers commented:
@ysiraichi a stack with two separate PRs, with the `dl_device` and `copy` keywords in the second PR, sounds perfectly fine to me. Merging the changes in one go will be nice, but functionally those two keywords don't interact with `max_version`, so it'll be a clean split.

albanD commented:
Separate PR on the stack sounds ok.
Note that you won't be able to land both together though if the bottom PR's CI is red though, they must all pass CI

ysiraichi committed:
Update

[ghstack-poisoned]

leofang commented:
Gentle ping @ysiraichi, any chance we can get this work wrapped up in the near future? Note that we just tagged DLPack v1.1: https://github.com/dmlc/dlpack/releases/tag/v1.1, which added a few more dtype enums.

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi commented:
Here is a summary of the latest updates after the last reviews:

- Renamed the DLPack API, removing the `Unversioned` suffix
- Renamed the DLPack API, adding `Versioned` to the newly added functions
- Removed the import of `to_dlpack_unversioned`
    - Using `torch._C._to_dlpack()` and `torch._C._to_dlpack_versioned()` directly
- Added `test_max_version.py`, which tests both the versioned and unversioned execution paths

@albanD @rgommers 
Sorry for the delay on this PR.
Let me know what you think of this new set of changes.

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi commented:
Hi, @albanD @rgommers. Sorry for taking too much time on this. I believe this stack should be good for another round of reviews. Let me know what you think.

ysiraichi committed:
Update

[ghstack-poisoned]

msaroufim commented on torch/utils/dlpack.py:
```diff
@@ -104,24 +106,34 @@ def from_dlpack(ext_tensor: Any) -> 'torch.Tensor':
+            dlpack = ext_tensor.__dlpack__(**kwargs)
+        except TypeError:
+            # If that doesn't work, try removing the `max_version` argument.
+            kwargs.pop("max_version")
```
why should we only pop max_version and not the other new kwargs?

msaroufim commented on torch/_tensor.py:
```diff
@@ -1690,9 +1690,14 @@ def __dlpack__(self, stream=None):
             If 1 (on CUDA) or 0 (on ROCM) then the default stream is used for
             synchronization.
+
+            max_version (tuple[int, int] or None): An optional Python tuple with
```
whats the tradeoff with being conservative on the default value?

ysiraichi replied:
That's correct.

msaroufim commented on torch/utils/dlpack.py:
```diff
@@ -3,28 +3,30 @@
     # Enums as in DLPack specification (aten/src/ATen/dlpack.h)
     kDLCPU = 1,
-    kDLGPU = 2,
-    kDLCPUPinned = 3,
```
why change the name?

ysiraichi replied:
This change reflects on the `aten/src/ATen/dlpack.h` enum.

albanD commented on torch/utils/dlpack.py:
```diff
@@ -3,28 +3,30 @@
     kDLCPU = 1,
-    kDLGPU = 2,
-    kDLCPUPinned = 3,
+    kDLCUDA = 2,
```
These are BC-breaking right?

ysiraichi replied:
Yes. But, this is needed if we want to follow the enum in `aten/src/ATen/dlpack.h`. This change was actually supposed to happen in #65047 (when the enum names actually changed).

albanD commented on torch/utils/dlpack.py:
```diff
@@ -104,24 +106,34 @@ def from_dlpack(ext_tensor: Any) -> 'torch.Tensor':
+        try:
+            # Try running __dlpack__ while specifying `max_version` argument.
+            dlpack = ext_tensor.__dlpack__(**kwargs)
+        except TypeError:
```
According to the standard, it says BufferError should be raised if you don't get a good max_version. Should we handle that here?

albanD reviewed (commented):
Sounds good!
Mostly small comments. 
The main thing would be to add a BC-breaking note on exactly which user facing APIs are changed.

ysiraichi commented on torch/_tensor.py:
```diff
@@ -1690,9 +1690,14 @@ def __dlpack__(self, stream=None):
         if has_torch_function_unary(self):
-            return handle_torch_function(Tensor.__dlpack__, (self,), self, stream)
+            args = (self, stream, max_version)
+            return handle_torch_function(Tensor.__dlpack__, (self,), *args)
```
Just checking if I understood your point correctly: we should add a check for `is_neg()` below, since there's already a `is_conj()` check, correct?

ysiraichi commented on test/test_dlpack.py:
```diff
@@ -281,6 +286,25 @@ def test_automatically_select_in_creation(self, device):
+        def test(device, **kwargs):
+            inp = make_tensor((5,), dtype=torch.float32, device=device)
+            out = torch.from_dlpack(inp.__dlpack__(**kwargs))
+            self.assertEqual(inp, out)
```
I don't think so. But I can check whether we are generating a (un)versioned DLPack tensor.

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

albanD reviewed (approved):
Ok!

atalman commented:
I believe this needs to be imported internally and landed internally since we are missing some internal changes. Error is during compilation, I believe there is a code thats using older constructs:
```
error: unknown type name 'DLManagedTensorVersioned'
   15 | TORCH_API DLManagedTensorVersioned* toDLPackVersioned(const Tensor& src);
...
: error: unknown type name 'DLManagedTensorVersioned'
   19 |     DLManagedTensorVersioned* src,
      |     ^
....
rror: use of undeclared identifier 'DLManagedTensorVersioned'
   52 | struct DLPackTraits<DLManagedTensorVersioned> {
      |      
```

albanD commented:
This is an executorch build issue for some reason.
@mergennachin do you know what is causing this?

ezyang commented:
LLM proposes

```
diff --git a/fbcode/caffe2/aten/src/ATen/dlpack.h b/fbcode/caffe2/aten/src/ATen/dlpack.h
--- a/fbcode/caffe2/aten/src/ATen/dlpack.h
+++ b/fbcode/caffe2/aten/src/ATen/dlpack.h
@@ -292,7 +292,7 @@
  *
  * \note This is the current standard DLPack exchange data structure.
  */
-struct DLManagedTensorVersioned {
+typedef struct DLManagedTensorVersioned {
   /*!
    * \brief The API and ABI version of the current managed Tensor
    */
@@ -326,7 +326,7 @@
   uint64_t flags;
   /*! \brief DLTensor which is being memory managed */
   DLTensor dl_tensor;
-};
+} DLManagedTensorVersioned;
 
 #ifdef __cplusplus
 }  // DLPACK_EXTERN_C
 ```

Which is kind of sus, not sure if this fixes the problem.

ysiraichi committed:
Update on "Upgrade to DLPack 1.0."


This PR makes the necessary changes in order to upgrade PyTorch DLPack
support to version 1.0. In summary, we add support for the following:

- Support both `DLManagedTensor` and `DLManagedTensorVersioned` when
  producing and consuming DLPack capsules
- New parameter for `__dlpack__` method: `max_version`
- Version checks:
    - Fallback to old implementation if no `max_version` or if version
      lower than 1.0
    - Check that the to-be-consumed capsule is of version up to 1.X

In order to accommodate these new specifications, this PR adds the
following main changes:

- `torch._C._to_dlpack_versioned` Python API (Module.cpp): new Python
API for creating a versioned DLPack capsule (called by `__dlpack__`
method)
- `DLPackTraits<T>` class (DLConvertor.h): select the correct
traits (e.g. capsule name, conversion functions) depending on which 
DLPack tensor class is being used
- `toDLPackImpl<T>` function (DLConvertor.cpp): populates the
common fields of both classes
- `fromDLPackImpl<T>` function (DLConvertor.cpp): constructs a tensor
from a DLPAck capsule
- `fillVersion<T>` function (DLConvertor.cpp): populates the version
field for `DLManagedTensorVersioned` (no-op for `DLManagedTensor`)
- `tensor_fromDLPackImpl<T>` function (tensor_new.cpp): outer function
for constructing a tensor out of a DLPack capsule that also marks the
capsule as used

cc ezyang gchanan

[ghstack-poisoned]

albanD reviewed (approved):
Thanks!
I don't have a devserver on hand, so trying to go with the regular land!

albanD commented:
@pytorchbot merge

albanD commented:
This landed smoothly, thank @ysiraichi for the change. We can merge the rest of the stack now!

tqchen commented:
great that this PR landed, just want to followup with https://github.com/pytorch/pytorch/pull/145000#issuecomment-2715050948 that we have 1.1 that comes with F8 and F4 data types, would be great to get help on landing support for them, hopefully as a matter of updating 

- https://github.com/pytorch/pytorch/blob/0e45023cf9cbe1cf18279c1b0d391ea9464e7731/aten/src/ATen/DLConvertor.cpp#L66-L90
- to map to https://github.com/dmlc/dlpack/blob/main/include/dlpack/dlpack.h#L162

ysiraichi commented:
I'm sorry, but I'm not able to work on this anymore. Feel free to pick this up.
